torch/csrc/autograd/profiler.h

```diff
@@ -131,15 +156,51 @@ struct TORCH_API Event final {
       uint16_t thread_id,
       bool record_cuda,
       at::RecordFunctionHandle handle = 0,
-      std::vector<std::vector<int64_t>>&& shapes = {})
+      std::vector<std::vector<int64_t>>&& shapes = {}, int node_id = -1, double relative_elapsed_cpu_us = -1)
       : name_(std::move(name)),
         kind_(kind),
         thread_id_(thread_id),
         handle_(handle),
-        shapes_(shapes) {
+        shapes_(shapes), node_id_(node_id), relative_elapsed_cpu_us_(relative_elapsed_cpu_us) {
     record(record_cuda);
   }
 
+  // Returns IValues corresponding to event structure, to be used for
+  // serialization.
+  at::IValue toIValues() const {
+    c10::impl::GenericList eventIValueList(at::AnyType::get());
+    eventIValueList.push_back(at::IValue(static_cast<int64_t>(kind_)));
+    eventIValueList.push_back(at::IValue(std::string(name_.str())));
+    eventIValueList.push_back(at::IValue(thread_id_));
+    eventIValueList.push_back(at::IValue(static_cast<double>(handle_)));
+    eventIValueList.push_back(at::IValue(node_id_));
+    eventIValueList.push_back(at::IValue(relative_elapsed_cpu_us_));
+    // CPU memory usage
+    eventIValueList.push_back(at::IValue(cpu_memory_usage_));
+    return at::IValue(eventIValueList);
+  }
+
+  // Reconstructs an event from IValues given by toIValues.
+  static Event fromIValue(at::IValue eventIValue) {
+    TORCH_INTERNAL_ASSERT(
+        eventIValue.isList(),
+        "Expected IValue to contain type c10::impl::GenericList");
+    auto ivalues = eventIValue.toList();
+
+    Event evt(
+      static_cast<EventKind>(ivalues.get(0).toInt()),
+      at::StringView(ivalues.get(1).toStringRef()),
+      ivalues.get(2).toInt(),
+      false, // TODO: record_cuda
+      static_cast<at::RecordFunctionHandle>(ivalues.get(3).toDouble()),
+      {}, // TODO: record shapes
+      ivalues.get(4).toInt(),
+      ivalues.get(5).toDouble() // relative time difference computed by remote end.
+    );
+    evt = evt.setCPUMemoryUsage(ivalues.get(6).toInt());
```

Comment:
Not sure if we should add a bunch of fields that should be set on the event when deserializing (such as node ID, time offset, cpu_memory_usage_, etc) into the constructor, or prefer something like the builder pattern also used here. It should be one or the other probably - open to feedback about which is preferred.